File: runner-ios/FleetRunner/Telemetry.swift
import Foundation
import Network
import UIKit

/// A long-lived `NWPathMonitor`, because there is no way to ask iOS "what is
/// carrying my traffic right now?" without one — the answer only arrives
/// through a monitor that has already been started.
///
/// Started once and kept, rather than spun up per request: a freshly started
/// monitor has no path yet, so a per-request one would report "unknown" for
/// exactly the first row of every vantage run.
private final class NetworkWatcher: @unchecked Sendable {
    static let shared = NetworkWatcher()
    private let monitor = NWPathMonitor()

    private init() {
        monitor.pathUpdateHandler = { _ in }
        monitor.start(queue: DispatchQueue(label: "com.taylab.fleetrunner.network-watcher"))
    }

    /// wifi / cellular / ethernet / unknown, the same four words the Android
    /// runner reports.
    ///
    /// Wi-Fi is checked before cellular so a phone holding both answers the way
    /// its traffic will actually go, which is also how Android's transport
    /// check reads. "unknown" is a word, not a guess: it covers no active path,
    /// a transport neither platform names, and the moment before the monitor
    /// has heard anything.
    func current() -> String {
        let path = monitor.currentPath
        if path.usesInterfaceType(.wifi) { return "wifi" }
        if path.usesInterfaceType(.cellular) { return "cellular" }
        if path.usesInterfaceType(.wiredEthernet) { return "ethernet" }
        return "unknown"
    }
}

enum Telemetry {

    /// What is carrying this device's traffic right now.
    ///
    /// Read per request during a vantage run rather than once per job: a device
    /// can leave wifi mid-run, and the rows on either side of that are honestly
    /// different measurements.
    static func networkType() -> String { NetworkWatcher.shared.current() }


    static func descriptor() -> DeviceDescriptor {
        var uts = utsname()
        uname(&uts)
        let machine = withUnsafeBytes(of: &uts.machine) { raw in
            String(decoding: raw.prefix(while: { $0 != 0 }), as: UTF8.self)
        }
        return DeviceDescriptor(
            model: machine,
            soc: machine,
            ramMb: Int64(ProcessInfo.processInfo.physicalMemory / (1024 * 1024)),
            os: "\(Self.platform)-\(UIDevice.current.systemVersion)",
            appVer: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0",
            platform: Self.platform,
            kind: Self.formFactor()
        )
    }

    /// Which Apple platform this binary was built for.
    ///
    /// Compile-time, because that is what the question is about: the same
    /// source produces an iOS app, a tvOS app and a watchOS app, and the
    /// numbers they report are only comparable when the table can tell them
    /// apart. `os` is built from this too, so a tvOS runner reports
    /// `tvos-18.2` rather than calling itself iOS with a television's memory.
    ///
    /// visionOS is checked before iOS: a "Designed for iPad" build running on
    /// Vision Pro compiles under `os(iOS)` and would otherwise report itself
    /// as an iPhone with a very unusual amount of RAM.
    static var platform: String {
        #if os(visionOS)
        return "visionos"
        #elseif os(tvOS)
        return "tvos"
        #elseif os(watchOS)
        return "watchos"
        #elseif targetEnvironment(macCatalyst)
        return "catalyst"
        #else
        return "ios"
        #endif
    }

    /// The shape, from the interface idiom the OS hands us.
    ///
    /// `.pad` covers both a real iPad and a "Designed for iPad" build on a
    /// headset, so visionOS is answered from the compile-time platform first —
    /// the same ordering, and the same reason, as `platform` itself.
    static func formFactor() -> String {
        #if os(visionOS)
        return "headset"
        #elseif os(watchOS)
        return "watch"
        #elseif os(tvOS)
        return "tv"
        #elseif targetEnvironment(macCatalyst)
        return "desktop"
        #else
        switch UIDevice.current.userInterfaceIdiom {
        case .pad: return "tablet"
        case .tv: return "tv"
        case .mac: return "desktop"
        case .vision: return "headset"
        default: return "phone"
        }
        #endif
    }

    /// Simulators report battery level -1; treat as 100 so constraints behave.
    static func batteryPct() -> Int {
        UIDevice.current.isBatteryMonitoringEnabled = true
        let level = UIDevice.current.batteryLevel
        return level < 0 ? 100 : Int(level * 100)
    }

    static func isCharging() -> Bool {
        UIDevice.current.isBatteryMonitoringEnabled = true
        switch UIDevice.current.batteryState {
        case .charging, .full: return true
        default: return false
        }
    }

    /// Shared thermal enum: nominal / fair / serious / critical.
    static func thermal() -> String {
        switch ProcessInfo.processInfo.thermalState {
        case .nominal: return "nominal"
        case .fair: return "fair"
        case .serious: return "serious"
        case .critical: return "critical"
        @unknown default: return "critical"
        }
    }

    /// phys_footprint in MB — labeled "phys_footprint" in results, never
    /// compared to Android PSS.
    static func physFootprintMb() -> Int64 {
        var info = task_vm_info_data_t()
        var count = mach_msg_type_number_t(
            MemoryLayout<task_vm_info_data_t>.size / MemoryLayout<integer_t>.size)
        let kr = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                task_info(mach_task_self_, task_flavor_t(TASK_VM_INFO), $0, &count)
            }
        }
        guard kr == KERN_SUCCESS else { return 0 }
        return Int64(info.phys_footprint) / (1024 * 1024)
    }

    static func beacon() -> BeaconSample {
        BeaconSample(batteryPct: batteryPct(), charging: isCharging(), thermal: thermal())
    }
}

File: runner-android/app/src/main/java/com/taylab/fleetrunner/telemetry/Telemetry.kt
package com.taylab.fleetrunner.telemetry

import android.app.ActivityManager
import android.app.UiModeManager
import android.content.Context
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.os.BatteryManager
import android.os.Build
import android.os.Debug
import android.os.PowerManager
import com.taylab.fleetrunner.BuildConfig
import com.taylab.fleetrunner.protocol.BeaconSample
import com.taylab.fleetrunner.protocol.DeviceDescriptor

object Telemetry {

    fun descriptor(context: Context): DeviceDescriptor {
        val am = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val mem = ActivityManager.MemoryInfo().also { am.getMemoryInfo(it) }
        val soc = if (Build.VERSION.SDK_INT >= 31) Build.SOC_MODEL else Build.HARDWARE
        return DeviceDescriptor(
            model = Build.MODEL,
            soc = soc,
            ramMb = mem.totalMem / (1024 * 1024),
            os = "android-${Build.VERSION.RELEASE}",
            appVer = BuildConfig.VERSION_NAME,
            kind = formFactor(context),
        )
    }

    /**
     * What shape of thing this APK is running on.
     *
     * The same binary installs on a phone, a tablet, a TV stick, a watch and a
     * headset, and nothing in `Build` distinguishes them: a Fire TV reports a
     * model and an SoC exactly the way a handset does, so a fleet that reads
     * only `Build` puts a television in a table of phones.
     *
     * `UiModeManager` is the system's own answer and is asked first. It is
     * authoritative for the four cases it names, and it is the field the
     * launcher itself routes on.
     *
     * A headset is asked about separately, and before the TV check, because
     * Quest reports UI_MODE_TYPE_NORMAL — it presents as an ordinary Android
     * phone environment — and only the VR feature flag tells the truth. The
     * feature constant is compared by string rather than by symbol so this
     * still compiles against an SDK that predates it.
     *
     * Phone versus tablet is the one case with no system answer at all, so it
     * falls back to the 600dp rule the resource system itself uses for
     * `sw600dp`. That keeps this agreeing with which layout the OS chose.
     */
    fun formFactor(context: Context): String {
        val pm = context.packageManager
        if (pm.hasSystemFeature(PackageManager.FEATURE_WATCH)) return "watch"
        // FEATURE_VR_HEADSET landed in API 26; the literal is what that constant
        // holds, and comparing the string keeps minSdk 24 compiling.
        if (pm.hasSystemFeature("android.hardware.vr.headset")) return "headset"

        val ui = context.getSystemService(Context.UI_MODE_SERVICE) as? UiModeManager
        when (ui?.currentModeType) {
            Configuration.UI_MODE_TYPE_TELEVISION -> return "tv"
            Configuration.UI_MODE_TYPE_WATCH -> return "watch"
            Configuration.UI_MODE_TYPE_CAR -> return "automotive"
            Configuration.UI_MODE_TYPE_APPLIANCE -> return "appliance"
        }
        // A TV that does not set the UI mode still declares the leanback
        // feature, which is what its launcher requires to show the app at all.
        if (pm.hasSystemFeature(PackageManager.FEATURE_LEANBACK)) return "tv"

        // No system call answers phone-versus-tablet, so use the rule the
        // resource system uses: sw600dp is where Android itself switches.
        return if (context.resources.configuration.smallestScreenWidthDp >= 600) "tablet" else "phone"
    }

    fun batteryPct(context: Context): Int {
        val bm = context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        return bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
    }

    fun isCharging(context: Context): Boolean {
        val bm = context.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        return bm.isCharging
    }

    /** Shared thermal enum: nominal / fair / serious / critical. Pre-API-29 has no signal → nominal. */
    fun thermal(context: Context): String {
        if (Build.VERSION.SDK_INT < 29) return "nominal"
        val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
        return when (pm.currentThermalStatus) {
            PowerManager.THERMAL_STATUS_NONE, PowerManager.THERMAL_STATUS_LIGHT -> "nominal"
            PowerManager.THERMAL_STATUS_MODERATE -> "fair"
            PowerManager.THERMAL_STATUS_SEVERE -> "serious"
            else -> "critical"
        }
    }

    /**
     * What is carrying this device's traffic right now: wifi / cellular /
     * ethernet / unknown, the same four words the iOS runner reports.
     *
     * Recorded on every vantage row rather than derived later, because the
     * whole point of measuring from where the device happens to be is that a
     * phone on cellular and a desktop on fibre give different answers — and a
     * timing with no idea what carried it is not comparable to anything. The
     * reading is taken per request: a device can leave wifi mid-job, and the
     * rows on either side of that are honestly different measurements.
     *
     * "unknown" covers the cases where the OS will not say — no active
     * network, a VPN or a transport we do not name — and is deliberately a
     * word rather than a guess.
     */
    fun networkType(context: Context): String {
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE)
            as? android.net.ConnectivityManager ?: return "unknown"
        val caps = try {
            cm.getNetworkCapabilities(cm.activeNetwork) ?: return "unknown"
        } catch (_: SecurityException) {
            // ACCESS_NETWORK_STATE denied: say so rather than invent a link.
            return "unknown"
        }
        return when {
            caps.hasTransport(android.net.NetworkCapabilities.TRANSPORT_WIFI) -> "wifi"
            caps.hasTransport(android.net.NetworkCapabilities.TRANSPORT_CELLULAR) -> "cellular"
            caps.hasTransport(android.net.NetworkCapabilities.TRANSPORT_ETHERNET) -> "ethernet"
            else -> "unknown"
        }
    }

    /** PSS in MB — labeled "pss" in results, never compared to iOS phys_footprint. */
    fun pssMb(): Long {
        val info = Debug.MemoryInfo().also { Debug.getMemoryInfo(it) }
        return info.totalPss / 1024L
    }

    fun beacon(context: Context): BeaconSample = BeaconSample(
        batteryPct = batteryPct(context),
        charging = isCharging(context),
        thermal = thermal(context),
    )
}
